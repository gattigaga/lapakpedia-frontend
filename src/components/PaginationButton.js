import React, { Component, createRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TweenLite, Power2 } from "gsap";
import { Link } from "react-router-dom";

import { capitalize } from "helpers/formatter";

const Container = styled.svg`
  width: 80px;
  height: 36px;
  cursor: ${props => (props.isDisabled ? "not-allowed" : "pointer")};
`;

const Layer = styled.rect`
  width: 100%;
  height: 100%;
  fill: ${props => (props.isDisabled ? "white" : "black")};
  stroke-width: 1px;
  stroke: black;
`;

const Label = styled.text`
  font-family: Roboto;
  font-size: 12px;
  letter-spacing: 0.1em;
  fill: ${props => (props.isDisabled ? "black" : "white")};
  text-anchor: middle;
  alignment-baseline: middle;
`;

const Line = styled.line`
  stroke: white;
  stroke-width: 2px;
  opacity: 0;
`;

class PaginationButton extends Component {
  constructor(props) {
    super(props);

    this.lines = [...Array(3)].map(createRef);
    this.label = createRef();

    this.handleAnimation = this.handleAnimation.bind(this);
    this.createLines = this.createLines.bind(this);
  }

  /**
   * Handle animation while hovered
   *
   * @param {boolean} isEnter - Check if mouse entering this component
   * @memberof PaginationButton
   */
  handleAnimation(isEnter) {
    const { type } = this.props;
    let labelX = 0;
    let linesXY = [];

    if (type === "prev") {
      labelX = isEnter ? "65%" : "50%";
      linesXY = [
        {
          x: isEnter ? "13%" : "35%",
          y: isEnter ? "50%" : "35%"
        },
        {
          x: isEnter ? "13%" : "35%"
        },
        {
          x: isEnter ? "13%" : "35%",
          y: isEnter ? "50%" : "65%"
        }
      ];
    } else {
      labelX = isEnter ? "35%" : "50%";
      linesXY = [
        {
          x: isEnter ? "87%" : "65%",
          y: isEnter ? "50%" : "35%"
        },
        {
          x: isEnter ? "87%" : "65%"
        },
        {
          x: isEnter ? "87%" : "65%",
          y: isEnter ? "50%" : "65%"
        }
      ];
    }

    TweenLite.to(this.label.current, 0.3, {
      ease: Power2.easeInOut,
      attr: {
        x: labelX
      }
    });

    TweenLite.to(this.lines[0].current, 0.3, {
      ease: Power2.easeInOut,
      opacity: isEnter ? 1 : 0,
      attr: {
        x1: linesXY[0].x,
        y1: linesXY[0].y
      }
    });

    TweenLite.to(this.lines[1].current, 0.3, {
      ease: Power2.easeInOut,
      opacity: isEnter ? 1 : 0,
      attr: {
        x1: linesXY[1].x
      }
    });

    TweenLite.to(this.lines[2].current, 0.3, {
      ease: Power2.easeInOut,
      opacity: isEnter ? 1 : 0,
      attr: {
        x1: linesXY[2].x,
        y1: linesXY[2].y
      }
    });
  }

  /**
   * Construct arrow by direction type
   *
   * @returns
   * @memberof PaginationButton
   */
  createLines() {
    const { type } = this.props;

    if (type === "prev") {
      return [
        <Line
          key={0}
          x1="13%"
          y1="50%"
          x2="20%"
          y2="35%"
          innerRef={this.lines[0]}
        />,
        <Line
          key={1}
          x1="13%"
          y1="50%"
          x2="35%"
          y2="50%"
          innerRef={this.lines[1]}
        />,
        <Line
          key={2}
          x1="13%"
          y1="50%"
          x2="20%"
          y2="65%"
          innerRef={this.lines[2]}
        />
      ];
    }

    return [
      <Line
        key={0}
        x1="87%"
        y1="50%"
        x2="80%"
        y2="35%"
        innerRef={this.lines[0]}
      />,
      <Line
        key={1}
        x1="87%"
        y1="50%"
        x2="65%"
        y2="50%"
        innerRef={this.lines[1]}
      />,
      <Line
        key={2}
        x1="87%"
        y1="50%"
        x2="80%"
        y2="65%"
        innerRef={this.lines[2]}
      />
    ];
  }

  render() {
    const { isDisabled, type, href } = this.props;

    return (
      <Link to={isDisabled ? "#" : href}>
        <Container
          isDisabled={isDisabled}
          onMouseEnter={() => !isDisabled && this.handleAnimation(true)}
          onMouseLeave={() => !isDisabled && this.handleAnimation(false)}
        >
          <Layer isDisabled={isDisabled} />
          {this.createLines()}
          <Label x="50%" y="52%" innerRef={this.label} isDisabled={isDisabled}>
            {capitalize(type)}
          </Label>
        </Container>
      </Link>
    );
  }
}

PaginationButton.propTypes = {
  isDisabled: PropTypes.bool,
  type: PropTypes.oneOf(["next", "prev"]),
  href: PropTypes.string
};

PaginationButton.defaultProps = {
  type: "next",
  href: "#"
};

export default PaginationButton;
