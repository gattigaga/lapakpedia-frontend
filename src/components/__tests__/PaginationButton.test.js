import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import PaginationButton from "components/PaginationButton";

describe("PaginationButton", () => {
  const setup = propOverrides => {
    const props = {
      ...propOverrides
    };

    const wrapper = shallow(<PaginationButton {...props} />);

    return {
      props,
      wrapper
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should renders in type 'prev'", () => {
    const { wrapper } = setup({ type: "prev" });

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should renders as disabled", () => {
    const { wrapper } = setup({ isDisabled: true });

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should render with link", () => {
    const { wrapper } = setup({ href: "https://github.com" });

    expect(wrapper.find("Link").props().to).toEqual("https://github.com");
  });
});
