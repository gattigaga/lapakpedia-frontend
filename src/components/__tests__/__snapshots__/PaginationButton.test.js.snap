// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationButton should renders as disabled 1`] = `
<Link
  replace={false}
  to="#"
>
  <styled.svg
    isDisabled={true}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <styled.rect
      isDisabled={true}
    />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="0"
      x1="87%"
      x2="80%"
      y1="50%"
      y2="35%"
    />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="1"
      x1="87%"
      x2="65%"
      y1="50%"
      y2="50%"
    />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="2"
      x1="87%"
      x2="80%"
      y1="50%"
      y2="65%"
    />
    <styled.text
      innerRef={
        Object {
          "current": null,
        }
      }
      isDisabled={true}
      x="50%"
      y="52%"
    >
      Next
    </styled.text>
  </styled.svg>
</Link>
`;

exports[`PaginationButton should renders default 1`] = `
<Link
  replace={false}
  to="#"
>
  <styled.svg
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <styled.rect />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="0"
      x1="87%"
      x2="80%"
      y1="50%"
      y2="35%"
    />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="1"
      x1="87%"
      x2="65%"
      y1="50%"
      y2="50%"
    />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="2"
      x1="87%"
      x2="80%"
      y1="50%"
      y2="65%"
    />
    <styled.text
      innerRef={
        Object {
          "current": null,
        }
      }
      x="50%"
      y="52%"
    >
      Next
    </styled.text>
  </styled.svg>
</Link>
`;

exports[`PaginationButton should renders in type 'prev' 1`] = `
<Link
  replace={false}
  to="#"
>
  <styled.svg
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <styled.rect />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="0"
      x1="13%"
      x2="20%"
      y1="50%"
      y2="35%"
    />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="1"
      x1="13%"
      x2="35%"
      y1="50%"
      y2="50%"
    />
    <styled.line
      innerRef={
        Object {
          "current": null,
        }
      }
      key="2"
      x1="13%"
      x2="20%"
      y1="50%"
      y2="65%"
    />
    <styled.text
      innerRef={
        Object {
          "current": null,
        }
      }
      x="50%"
      y="52%"
    >
      Prev
    </styled.text>
  </styled.svg>
</Link>
`;
